File: tests/bench/04_strings.go
// Benchmark 04: Strings — concatenation, format, split/join
// Measures string allocation, formatting overhead, string ops
package main

import (
	"fmt"
	"strings"
	"time"
)

func main() {
	// Part A: Concatenation
	nConcat := 20_000_000
	start := time.Now()
	var builder strings.Builder
	for i := 0; i < nConcat; i++ {
		builder.WriteByte('a')
	}
	s := builder.String()
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("[bench] string_concat: %d iterations in %d ms (len=%d)\n", nConcat, elapsed, len(s))

	// Part B: Interpolation (Sprintf)
	nInterp := 20_000_000
	start = time.Now()
	last := ""
	for i := 0; i < nInterp; i++ {
		last = fmt.Sprintf("%d", i)
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] string_interp: %d iterations in %d ms (last=%s)\n", nInterp, elapsed, last)

	// Part C: Split + Join
	base := "the quick brown fox jumps over the lazy dog"
	nSplit := 1_000_000
	start = time.Now()
	result := ""
	for i := 0; i < nSplit; i++ {
		parts := strings.Split(base, " ")
		result = strings.Join(parts, "-")
	}
	elapsed = time.Since(start).Milliseconds()
	fmt.Printf("[bench] string_split_join: %d iterations in %d ms (result=%s)\n", nSplit, elapsed, result)
}

File: tests/bench/04_strings.rs
// Benchmark 04: Strings — concatenation, format, split/join
// Measures string allocation, formatting overhead, string ops

use std::time::Instant;

fn main() {
    // Part A: Concatenation
    let n_concat = 20_000_000;
    let start = Instant::now();
    let mut s = String::new();
    for _ in 0..n_concat {
        s.push('a');
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] string_concat: {} iterations in {} ms (len={})", n_concat, elapsed, s.len());

    // Part B: Interpolation (format!)
    let n_interp = 20_000_000;
    let start = Instant::now();
    let mut last = String::new();
    for i in 0..n_interp {
        last = format!("{}", i);
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] string_interp: {} iterations in {} ms (last={})", n_interp, elapsed, last);

    // Part C: Split + Join
    let base = "the quick brown fox jumps over the lazy dog";
    let n_split = 1_000_000;
    let start = Instant::now();
    let mut result = String::new();
    for _ in 0..n_split {
        let parts: Vec<&str> = base.split(' ').collect();
        result = parts.join("-");
    }
    let elapsed = start.elapsed().as_millis();
    println!("[bench] string_split_join: {} iterations in {} ms (result={})", n_split, elapsed, result);
}
